### PR TITLE
MAILBOX-290 Should be normal attachment in case of inline but no cid

### PR DESCRIPTION
**Pull request by quynhn**

### CODE REVIEW

File: mailbox/api/src/test/java/org/apache/james/mailbox/model/MessageAttachmentTest.java [hunk 1]
```diff
@@ -71,17 +71,19 @@ public void buildShouldSetIsInlineDefaultValueWhenNotGiven() {
         assertThat(messageAttachment.isInline()).isFalse();
     }
 
-    @Test(expected=IllegalStateException.class)
-    public void buildShouldThrowWhenIsInlineAndNoCid() {
+    @Test
+    public void buildShouldChangeInlineWhenIsInlineAndNoCid() {
```
Comment (rouazana) [inline mailbox/api/src/test/java/org/apache/james/mailbox/model/MessageAttachmentTest.java]: s/ChangeInline/ChangeIsInline/

File: mailbox/api/src/main/java/org/apache/james/mailbox/model/MessageAttachment.java [hunk 2]
```diff
@@ -76,7 +76,7 @@ public MessageAttachment build() {
                 isInline = false;
             }
             if (isInline && !cid.isPresent()) {
-                throw new IllegalStateException("'cid' is mandatory for inline attachments");
+                isInline = false;
```
Comment (mbaechler) [inline mailbox/api/src/main/java/org/apache/james/mailbox/model/MessageAttachment.java]: there's something wrong : I can set isInline to true and have it set to false after build.
That probably means we are not using the right abstraction

Comment (chibenwa) [inline mailbox/api/src/main/java/org/apache/james/mailbox/model/MessageAttachment.java]: Optional on isInline

then : 

```
        public MessageAttachment build() {
            Preconditions.checkState(attachment != null, "'attachment' is mandatory");

            boolean computedIsInline = isInline.orElse(false) && cid.isPresent();
            return new MessageAttachment(attachment, name, cid, computedIsInline);
        }
```

File: mailbox/api/src/main/java/org/apache/james/mailbox/model/MessageAttachment.java [hunk 3]
```diff
@@ -75,9 +75,6 @@ public MessageAttachment build() {
             if (isInline == null) {
                 isInline = false;
```
Comment (chibenwa) [inline mailbox/api/src/main/java/org/apache/james/mailbox/model/MessageAttachment.java]: Please no variable reallocation!

isInline should be an Optional<Boolean> to ease computation.

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMessageManager.java [hunk 4]
```diff
@@ -456,11 +456,15 @@ private boolean hasNonInlinedAttachment(List<MessageAttachment> attachments) {
...
             @Override
             public boolean apply(MessageAttachment input) {
-                return !input.isInline();
+                return isNotRealInline(input);
             }
         });
     }
 
+    private boolean isNotRealInline(MessageAttachment input) {
+        return !input.isInline() || (input.isInline() && !input.getCid().isPresent());
```
Comment (chibenwa) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMessageManager.java]: Put more brackets, it's not understandable at all

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMessageManager.java [hunk 5]
```diff
@@ -456,11 +456,15 @@ private boolean hasNonInlinedAttachment(List<MessageAttachment> attachments) {
         return FluentIterable.from(attachments).anyMatch(new Predicate<MessageAttachment>() {
             @Override
             public boolean apply(MessageAttachment input) {
-                return !input.isInline();
+                return isNotRealInline(input);
             }
         });
     }
 
+    private boolean isNotRealInline(MessageAttachment input) {
```
Comment (chibenwa) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMessageManager.java]: isInlineWithoutCid

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMessageManager.java [hunk 6]
```diff
@@ -456,11 +456,15 @@ private boolean hasNonInlinedAttachment(List<MessageAttachment> attachments) {
         return FluentIterable.from(attachments).anyMatch(new Predicate<MessageAttachment>() {
```
Comment (chibenwa) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMessageManager.java]: You can extract this predicate in a constant for readability

File: mailbox/api/src/test/java/org/apache/james/mailbox/model/MessageAttachmentTest.java [hunk 7]
```diff
@@ -71,17 +71,19 @@ public void buildShouldSetIsInlineDefaultValueWhenNotGiven() {
         assertThat(messageAttachment.isInline()).isFalse();
     }
 
-    @Test(expected=IllegalStateException.class)
-    public void buildShouldThrowWhenIsInlineAndNoCid() {
+    @Test
+    public void buildShouldChangeIsInlineWhenIsInlineAndNoCid() {
```
Comment (rouazana) [inline mailbox/api/src/test/java/org/apache/james/mailbox/model/MessageAttachmentTest.java]: you can remove this test now (or rename it)

File: mailbox/api/src/main/java/org/apache/james/mailbox/model/MessageAttachment.java [hunk 8]
```diff
@@ -66,19 +66,13 @@ public Builder cid(Cid cid) {
         }
 
         public Builder isInline(boolean isInline) {
-            this.isInline = isInline;
+            this.isInline = Optional.fromNullable(isInline);
```
Comment (rouazana) [inline mailbox/api/src/main/java/org/apache/james/mailbox/model/MessageAttachment.java]: Optional.of as a boolean cannot be null

File: mailbox/api/src/test/java/org/apache/james/mailbox/model/MessageAttachmentTest.java (hunk 7 above)
Comment (quynhn) [inline mailbox/api/src/test/java/org/apache/james/mailbox/model/MessageAttachmentTest.java]: Yes, it does not make sense any more so I propose to delete it

File: mailbox/api/src/main/java/org/apache/james/mailbox/model/MessageAttachment.java (hunk 8 above)
Comment (quynhn) [inline mailbox/api/src/main/java/org/apache/james/mailbox/model/MessageAttachment.java]: Yes, and I change boolean to Boolean

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/model/Message.java [hunk 9]
```diff
@@ -244,11 +244,15 @@ protected static boolean areAttachedMessagesKeysInAttachments(ImmutableList<Atta
...
     private static boolean hasAttachment(List<Attachment> attachments) {
         return attachments.stream()
-                .filter(attachment -> !attachment.isIsInline())
+                .filter(attachment -> isNotRealInline(attachment))
                 .findAny()
                 .isPresent();
     }
 
+    private static boolean isNotRealInline(Attachment attachment) {
+        return !attachment.isIsInline() || (attachment.isIsInline() && !attachment.getCid().isPresent());
```
Comment (chibenwa) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/model/Message.java]: Hmmmmm

Still not readable yet.

I would propose : 

```
return !attachment.isIsInline() || isInlinedWithoutCid(attachment);
```

With : 

```
private static boolean isInlinedWithoutCid(Attachment attachment) {
    return attachment.isIsInline() && !attachment.getCid().isPresent();
}
```

Moreover **isNotRealInline** is a bad method name. And this is code duplication. Detecting if this is a "real attachment" should be a concern of Attachment class. Please move logic for this in a method called **isInlinedWithCid**, and write test for it.

**Pull request merged**